compiler/src/yul/mappers/expressions.rs

```diff
@@ -103,15 +101,40 @@ pub fn expr_call(
                 CallType::TypeConstructor { .. } => Ok(yul_args[0].to_owned()),
                 CallType::SelfAttribute { func_name } => {
                     let func_name = names::func_name(func_name);
-
                     Ok(expression! { [func_name]([yul_args...]) })
                 }
-                CallType::ValueAttribute { .. } => {
-                    if let fe::Expr::Attribute { value, .. } = &func.node {
-                        expr(context, value)
-                    } else {
-                        unreachable!()
+                CallType::ValueAttribute => {
+                    if let fe::Expr::Attribute { value, attr } = &func.node {
+                        let value_attributes =
+                            context.get_expression(value).expect("invalid attributes");
+
+                        return match (value_attributes.typ.to_owned(), attr.node) {
+                            (Type::Contract(contract), func_name) => {
+                                Ok(call_operations::contract_call(
+                                    contract,
+                                    func_name.to_owned(),
+                                    expr(context, value)?,
+                                    yul_args,
+                                ))
+                            }
+                            (_, func_name) => {
+                                match builtins::Method::from_str(func_name)
+                                    .expect("uncaught analyzer error")
+                                {
+                                    // Copying is done in `expr(..)` based on the move location set
+                                    // in the expression's attributes, so we just map the value for
+                                    // `to_mem` and `clone`.
+                                    builtins::Method::ToMem => expr(context, value),
+                                    builtins::Method::Clone => expr(context, value),
+                                    builtins::Method::AbiEncode => todo!(),
+                                    builtins::Method::AbiEncodePacked => todo!(),
+                                    builtins::Method::Keccak256 => todo!(),
+                                }
+                            }
+                        };
                     }
+
+                    panic!("invalid attributes")
```

Comment:
Was this deliberate?

Reply:
If the call type is `ValueAttribute` and the function expression doesn't actually consist of an attribute expression, then I think it makes sense to panic with "invalid attributes".

Reply:
Sorry, should have been more clear. What I meant was, why not `return CompileError::static_str("invalid attributes")` like we do at these other places. Now this is the only place in the mapper that does a straight panic (not counting `unimplemented!()`)

Reply:
I think we need to do some refactoring with error handling in the compiler library. In general, I think result returns should only be used when it's useful to the user. An error here means that the there is a bug in the compiler, so I think we should panic.

I wrote up #168 for this refactor.